### ISSUE TITLE
getmail keeps downloading the whole mailbox

### ISSUE THREAD

**Issue by zazy** [getmail6/getmail6#112]
I'm using getmail to create an mbox copy of a mailbox. The mailbox contains almost 20000 messages that uses almost 15GB of space.
The server seems to stop responding after a certain amount of time (maybe a security measure) so getmail ends with a timout at some point. For this reason I set read_all option to false so that getmail can restart from where it was before. Here is what happens:

- I launch getmail for the first time, it downloads almos 18000 messages before server stops responding. The mailbox file is 11GB.
- I launch getmail for the second time. I downloaded all 18000 messages again from the beginning. The mbox file is now almost 20 GB
- Launching getmail again will result in appending emails to the mbox file over and over

Command line is as simple as getmail -g.

**getmail --dump** result:
```
getmail version 6.18.6
Copyright (C) 1998-2021 Charles Cazabon and others. Licensed under GNU GPL version 2.
getmail configuration:
  getmail version 6.18.6
  Python version 3.8.10 (default, Nov 26 2021, 20:14:08) 
[GCC 9.3.0]
  retriever:  SimpleIMAPSSLRetriever(ca_certs="None", certfile="None", getmaildir=".", imap_on_delete="None", imap_search="None", keyfile="None", mailboxes="('ALL',)", move_on_delete="None", password="*", password_command="()", port="993", record_mailbox="True", server="outlook.office365.com", ssl_cert_hostname="None", ssl_ciphers="None", ssl_fingerprints="()", ssl_version="None", timeout="180", use_cram_md5="False", use_kerberos="False", use_peek="True", use_xoauth2="False", username="**********@*************.com")
  destination:  Mboxrd(locktype="lockf", path="./mbox", user="None")
  options:
    delete : False
    delete_after : 0
    delete_bigger_than : 0
    delivered_to : True
    fingerprint : False
    logfile : None
    max_bytes_per_session : 0
    max_message_size : 0
    max_messages_per_session : 0
    message_log : None
    message_log_syslog : False
    message_log_verbose : False
    netrc_file : None
    read_all : False
    received : True
    skip_imap_fetch_size : False
    use_netrc : False
    verbose : 1
```
and the **germailrc** file:

```
[retriever]
type = SimpleIMAPSSLRetriever
server = outlook.office365.com
port = 993
mailboxes = ALL
username = ***********@*********.com
password = **********

[destination]
type = Mboxrd
path = ./mbox

[options]
read_all = false
```

**Comment by rpuntaie** [comment]
If `delete` is false, the message will not be deleted from the server: the server will list the message again on the next try.
I think `read_all = true` and `delete=True` would be better. Use `max_message_per_session` to limit the amount of messages. You can use `imap_search` to select subsets of emails.

**Comment by emtiu** [comment]
OP says they wish to make "a copy" of the remote mailbox, so deleting is probably not desired.

This looks like the edge case where already downloaded mails are not recorded in outmail if the connection times out partway through the download.

**Comment by zazy** [comment]
@rpuntaie as @emtiu says delete messages is not desired. The `read_all` documentation says that if `read_all` is set to false gemail will not download already seen messages. I interpreted it like: "If a massage had been already downloaded it will not be downloaded again".

Seems the case pointed by @emtiu: when getmail times out (in my case the timeout error is "The read operation timed out") previously downloaded email are not recorded and they will be downloaded again. In this case there should be a better error management so that gathered emails can be indexed and don't let the job done be lost.

I will try @rpuntaie suggestion to use `max_messages_per_session` to see if it can be a workaround to build the whole mbox. Will let you know within few hours.

**Comment by rpuntaie** [comment]
Thanks for the heads-up.
In the code `oldmail-<server>-<port>-<user>-<mailbox>` seems to be written in all paths,
but `getmail` spawns subprocesses,
which probably get killed before completing the normal code path.
I added the `to_oldmail_on_each_mail` option.

**Comment by zazy** [comment]
Thank you @rpuntaie for adding that option. Sorry for relate in reporting but got very busy days.
I know this issue had been closed but I'd love to add resultings about tests.

I tested the workaround suggesting to use `max_message_per_session` and it worked. All messages had been successfully retrieved but this required to launch getmail lots of time.
I tested the `to_oldmail_on_each_mail` option and it worked flawlessly. The server stopped at some time, got the time-out and when launched again, getmail retrieved only missing messages. Launched two times got the whole mailbox. I didn't noticed any slowdown.

In both cases I got a single mbox file with all messages in it. This way rebuilding the folder structure is impossible. The additional X-getmail-retrieved-from-mailbox is not useful for this purpose because no email clients can handle it.

As a future enhancement there could be a parameter for command line to get the list of mailboxes from the server so that when you don't know how server names it's folder structure this list can be used with the param `mailboxes` for the IMAPs retrievers and rebuild the structure downloading messages in several mbox formatted files.
Even better getmail can create separated files to put messages into when `mailboxes = ALL` is specified.
